From 163e32065c1d498337f33d233f0b03ad26dcb35e Mon Sep 17 00:00:00 2001
From: Luiz Angelo Daros de Luca <luizluca@gmail.com>
Date: Wed, 11 Jan 2023 02:08:02 -0300
Subject: [PATCH] bitfield.h: add FIELD_WIDTH()

A field might have its bits stored at different non sequential
addresses. In order to assemble that value back, FIELD_WIDTH() returns
the number of bits in a mask for left-shifting the most significant
part.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>
---
 include/linux/bitfield.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/include/linux/bitfield.h
+++ b/include/linux/bitfield.h
@@ -89,6 +89,18 @@
 	})
 
 /**
+ * FIELD_WIDTH() - returns the number of bits in the field
+ * @_mask: shifted mask defining the field's length and position
+ *
+ * FIELD_WIDTH() returns the number of 1-bit specified by @_mask.
+ */
+#define FIELD_WIDTH(_mask)						\
+	({								\
+		__BF_FIELD_CHECK(_mask, 0ULL, 0ULL, "FIELD_MAX: ");	\
+		__bf_shf(~FIELD_MAX(_mask));				\
+	})
+
+/**
  * FIELD_FIT() - check if value fits in the field
  * @_mask: shifted mask defining the field's length and position
  * @_val:  value to test against the field
